### ISSUE TITLE
`EnumerableGenericStaticMethodDynamicMethodReturnTypeExtension` exception excludes `EloquentCollection` **children**

### ISSUE THREAD

**Issue by rudiedirkx** [nunomaduro/larastan#2089]
- Larastan Version: v2.9.9
- Laravel Version: v11.23.5

### Description

See my rant in https://github.com/larastan/larastan/discussions/2088#discussioncomment-11044323 😆 

There is a partial exception in `EnumerableGenericStaticMethodDynamicMethodReturnTypeExtension` for Eloquent collections, because they're more strictly genericized or whatever you call that. So Larastan doesn't try to be smart about `groupBy()`. But Larastan ONLY doesn't try to be smart for the exact `EloquentCollection` class, not custom collections that descend from that class, like my `App\Base\ModelCollection` and `App\Collections\UserCollection`.

In practice this isn't often a bug, because the extension returns what Laravel does anyway. For `groupBy()` that is `@return static<array-key, static<array-key, TValue>>`, with 2x `static`, meaning you can make a `UserCollection` of `UserCollections`. My `ModelCollection` overwrites `groupBy()` to return a `Illuminate\Support\Collection` of `UserCollection`:

```php
/** @return BaseCollection<array-key, $this> */
public function groupBy($groupBy, $preserveKeys = false) {
	$groups = parent::groupBy($groupBy, $preserveKeys);
	return BaseCollection::make($groups->all());
}
```

but because of `EnumerableGenericStaticMethodDynamicMethodReturnTypeExtension` Phpstan still thinks it's a `UserCollection` of `UserCollection`.

If I hack Larastan to check it like this, it works:

```diff
- if ($methodReflection->getDeclaringClass()->getName() === EloquentCollection::class) {
+ $collectionClass = $methodReflection->getDeclaringClass()->getName();
+ if (is_a(new $collectionClass, EloquentCollection::class)) {
```

That's probably a ridiculous way to check a parent class, but you get it.

**Comment by calebdw** [comment]
You can't use `is_a` in phpstan because that's a runtime check that's not guaranteed to work

it should be:

```php
if ($methodReflection->getDeclaringClass()->is(EloquentCollection::class)) {
```

**Comment by rudiedirkx** [comment]
Yeah, I cheat like that sometimes, because I don't know Phpstan internals. Yours is even shorter and more readable 😆

**Comment by canvural** [comment]
PHPStan now has support for `static<>` and the extension we had here for the same case are removed. So, closing this issue. Please test it with newer versions of Larastan and PHPStan.
